packages/std/src/ibc/callbacks.rs

```diff
@@ -0,0 +1,173 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcDstCallback {
+    /// The destination chain address that should receive the callback.
+    pub address: String,
+    /// Optional gas limit for the callback (in Cosmos SDK gas units)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+/// The type of IBC source chain callback that is being called.
+///
+/// IBC source chain callbacks are needed for cases where your contract triggers the sending of an
+/// IBC packet through some other message (i.e. not through [`IbcMsg::SendPacket`]) and needs to
+/// know whether or not the packet was successfully received on the other chain.
+/// A prominent example is the [`IbcMsg::Transfer`] message. Without callbacks, you cannot know
+/// whether the transfer was successful or not.
+///
+/// Note that there are some prerequisites that need to be fulfilled to receive source chain callbacks:
+/// - The contract must implement the `ibc_source_chain_callback` entrypoint.
+/// - The module that sends the packet must be wrapped by an `IBCMiddleware`
+///   (i.e. the source chain needs to support callbacks for the message you are sending).
+/// - You have to add json-encoded [`IbcCallbackData`] to a specific field of the message.
+///   For `IbcMsg::Transfer`, this is the `memo` field.
+/// - The receiver of the callback must also be the sender of the message.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub enum IbcSourceChainCallbackMsg {
+    Acknowledgement(IbcPacketAckMsg),
+    Timeout(IbcPacketTimeoutMsg),
+}
+
+/// The message type of the IBC destination chain callback.
+///
+/// The IBC destination chain callback is needed for cases where someone triggers the sending of an
+/// IBC packet through some other message (i.e. not through [`IbcMsg::SendPacket`]) and
+/// your contract needs to know that it received this.
+/// The callback is called after the packet was successfully acknowledged on the destination chain.
+/// A prominent example is the [`IbcMsg::Transfer`] message. Without callbacks, you cannot know
+/// that someone sent you IBC coins.
+///
+/// Note that there are some prerequisites that need to be fulfilled to receive source chain callbacks:
+/// - The contract must implement the `ibc_destination_chain_callback` entrypoint.
+/// - The module that receives the packet must be wrapped by an `IBCMiddleware`
+///   (i.e. the destination chain needs to support callbacks for the message you are being sent).
+/// - You have to add json-encoded [`IbcCallbackData`] to a specific field of the message.
+///   For `IbcMsg::Transfer`, this is the `memo` field.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcDestinationChainCallbackMsg {
+    pub packet: IbcPacket,
+    pub ack: IbcAcknowledgement,
```

Comment:
On the destination chain, how do we know if there is an ack or not?
- What about notifications for timeouts (e.g. someone tried to send you funds but failed)?
- What about cases where a packet is received but no acknowledgement is written because the protocol never writes them?

Reply:
In general we don't know that, but I just learned that my understanding of how this works was slightly off. Here's how it actually works:
If the ack is async (using WriteAcknowledgement), then the callback is called both for successful and unsuccessful acknowledgements.
If the ack is not async (by returning, then the callback is only called if it was successful.

I originally thought only successful acks cause a callback, so we need to change the interface here to convey whether it was successful or not.

* Regarding timeout: The destination chain does not get to know about timeouts because (AFAIK) the IBC protocol has no mechanism for that. It would require the protocol built on top to support it (by sending another packet).
* Regarding no ack: If no ack is written, no callback is received.

Reply:
The destination chain doesn't know about timeouts at all (even core ibc), so it is not possible to do a callback anyway.

If `OnRecvPacket` returns an unsuccessful ack, core ibc reverts the cached context, so there is no reason to make a callback since it'd be reverted anyway. The intuition is that an unsuccessful ack means that the packet couldn't be received. The knowledge that a sender has attempted and failed to make a transfer is only useful to the sender (not the destination).

However, things get more complicated with async acknowledgements. (Note that no application in ibc-go currently supports async acknowledgements.) This is because async acknowledgements are initiated by an application module, not core ibc (unlike RecvPacket). This means that the decision to revert state on unsuccessful acknowledgement is up to the application module, and thus, we perform the callback in this case.